### PR TITLE
block id in vote could be empty

### PR DESCRIPTION
**Pull request by yihuang**
Somehow, the block id field in vote of precommits of block could also be empty.
Real world test case included.

### CODE REVIEW

File: tendermint/src/serializers.rs [hunk 1]
```diff
@@ -144,3 +144,41 @@ where
     let v = Option::deserialize(deserializer)?;
     Ok(v.map(|Wrapper(a)| a))
 }
+
+/// Parse empty block id as None.
+pub(crate) fn parse_non_empty_block_id<'de, D>(
```
Comment (ebuchman) [inline tendermint/src/serializers.rs]: I'm kind of hoping this will go away with #101. In which case maybe we'll want to use an explicit VoteValue enum with variants BlockId and Nil ...

File: tendermint/src/amino_types/vote.rs [hunk 2]
```diff
@@ -58,9 +58,9 @@ impl From<&vote::Vote> for Vote {
             vote_type: vote.vote_type.to_u32(),
             height: vote.height.value() as i64, // TODO potential overflow :-/
             round: vote.round as i64,
-            block_id: Some(BlockId {
-                hash: vote.block_id.hash.as_bytes().to_vec(),
-                parts_header: vote.block_id.parts.as_ref().map(PartsSetHeader::from),
+            block_id: vote.block_id.as_ref().map(|block_id| BlockId {
```
[the remark this answers is not in the document]
Comment (yihuang) [reply tendermint/src/amino_types/vote.rs]: As far as I know, the rust version of amino proto is like this:
* `None` is skiped, as the same as default value of protobuf.
* `Some(value)` encodes the value as is, won't skip it if it's the default value.
And the go version seems to automatically skip default values, so if rust version encodes `Some(default_value)`, it won't match the go version, we need to prevent this from happening.

File: tendermint/src/amino_types/vote.rs [hunk 3]
```diff
@@ -58,9 +58,9 @@ impl From<&vote::Vote> for Vote {
...
+                hash: block_id.hash.as_bytes().to_vec(),
+                parts_header: block_id.parts.as_ref().map(PartsSetHeader::from),
```
[the remark this answers is not in the document]
Comment (yihuang) [reply tendermint/src/amino_types/vote.rs]: I believe there are situation when block_id has valid `hash`, but don't has `parts_header`, or have a default `parts_header`(can't recall the case for now), for go version of protobuf, default value is skipped automatically when encoding, but rust version won't check the default value automatically, so we always needs an `Option` to handle this situation.

File: tendermint/src/serializers.rs (hunk 1 above)
Comment (yihuang) [reply tendermint/src/serializers.rs]: I actually don't think `enum` would help us in these cases, because it's adding a branch just like `Option`, you just move the branch from `Option` to your own type.
The core problem lies in the protobuf semantic of optional field and default values, and the inconsistency between json and binary version of protobuf encoding.
In the semantic of protobuf, basically, there's no separation between null value and default value, they can be encoded as the same(skip), but in the json version, the default value is encoded as is, not skipped or null, it is actually correct in protobuf, because a skipped or null value is the same as a default value, but when we use `Option` to distingish null value and default value differently, we need extra trouble.

**Comment by yihuang** [comment]
Let me rephrase our current situation here:
- The semantic of protobuf(also go-amino) don't distinguish empty values and default ones, and the json encoding doesn't even skip default values, for example:
```
type subMessage struct {
	Height int
	Name   string
}

type bcMessage struct {
	Sub subMessage
}

// With MarshalBinaryBare, bcMessage{}, bcMessage{Sub: subMessage{}}, bcMessage{Sub: subMessage{Height: 0, Name: ""}} all encodes to empty bytes.

// MarshalJson(bcMessage{}) -> {"type":"bcMessage","value":{"Sub":{"Height":"0","Name":""}}}
```

- According to my test cases with amino-rs, here's the behavior:
  - If you define `Option<i64>`, `Some(0)` is not encoded to empty.
  - message value with all fields set to default values, is not encoded to empty.
  - the json encoding is even messier when dealing with the empty/default values.
```
#[derive(Clone, PartialEq, Message)]
pub struct OptionalScalar {
    #[prost(int64, optional, tag = "1")]
    pub height: Option<i64>,
    #[prost(string, tag = "2")]
    pub name: String,
}

#[derive(Clone, PartialEq, Message)]
pub struct SubMessage {
    #[prost(int64, tag = "1")]
    pub height: i64,
    #[prost(string, tag = "2")]
    pub name: String,
}

#[derive(Clone, PartialEq, Message)]
pub struct TestMessage {
    #[prost(message, tag = "1")]
    pub sub: Option<SubMessage>,
}

fn check_amino_encoding() {
    let buf = AminoMessage::bytes_vec(&TestMessage {
        sub: Some(SubMessage {
            height: 0,
            name: "".to_owned(),
        }),
    });
    assert_ne!(buf, vec![]);

    assert_ne!(
        AminoMessage::bytes_vec(&OptionalScalar {
            height: Some(0),
            name: "".to_owned(),
        }),
        AminoMessage::bytes_vec(&OptionalScalar {
            height: None,
            name: "".to_owned(),
        })
    );
}
```

**Comment by Liamsi** [comment]
Can we somehow enable circleci for PRs from forks? That would be amazing!

**Comment by Liamsi** [comment]
merging in master reverted removing `parse_non_empty_block_id` from `header.rs`. Trying to fix this now. Not sure I can push to the fork to impact this PR but we will see shortly.
